crates/iota-names-indexer/src/main.rs

```diff
@@ -4,57 +4,167 @@
+use crate::{metrics::PrometheusServer, worker::run_iota_names_reader};
+
+// Define the `GIT_REVISION` and `VERSION` consts
+bin_version::bin_version!();
+
+#[derive(Parser)]
+#[command(
+    name = env!("CARGO_BIN_NAME"),
+    about = env!("CARGO_PKG_DESCRIPTION"),
+    author,
+    version = VERSION,
+    propagate_version = true,
+)]
+enum Command {
+    Start {
+        /// The URL of an IOTA node to get data from.
+        #[arg(long, default_value = "http://localhost:9000")]
+        node_url: String,
+        #[arg(long, default_value_t = 1)]
+        num_workers: usize,
+    },
+}
+
+impl Command {
+    async fn execute(self) -> Result<()> {
+        match self {
+            Command::Start {
+                node_url,
+                num_workers,
+            } => {
+                info!("Starting IOTA Names Indexer");
+
+                let prometheus = PrometheusServer::new();
+                let registry = prometheus.registry();
+
+                let cancel_token = CancellationToken::new();
+
+                let mut tasks: JoinSet<Result<()>> = JoinSet::new();
+
+                // Spawn the prometheus API
+                let handle = cancel_token.clone();
+                tasks.spawn(async move {
+                    prometheus.start(handle).await?;
+                    Ok(())
+                });
+
+                // Spawn the metrics worker
+                let handle = cancel_token.clone();
+                tasks.spawn(async move {
+                    let worker = IotaNamesWorker::new(
+                        IotaNamesConfig::from_env().unwrap_or_default(),
+                        Arc::new(IotaNamesMetrics::new(&registry)),
+                    );
+
+                    tokio::select! {
+                        res = run_iota_names_reader(worker, &node_url, &registry, handle.clone(), num_workers) => {res?;},
+                        _ = handle.cancelled() => {},
+                    }
+                    Ok(())
+                });
+
+                let mut exit_code = Ok(());
+
+                tokio::select! {
+                    res = interrupt_or_terminate() => {
+                        if let Err(err) = res {
+                            tracing::error!("subscribing to OS interrupt signals failed with error: {err}; shutting down");
+                            exit_code = Err(err);
+                        } else {
+                            tracing::info!("received interrupt; shutting down");
+                        }
+                    },
+                    res = tasks.join_next() => {
+                        if let Some(Ok(Err(err))) = res {
+                            tracing::error!("a worker failed with error: {err}");
+                            exit_code = Err(err);
+                        }
+                    },
+                }
+
+                cancel_token.cancel();
+
+                // Allow the user to abort if the tasks aren't shutting down quickly.
```

Comment:
Are we doing this in the other indexers?

Reply:
No, but it's good practice

Reply:
Feels like there has to be a better way to do this but ok